### PR TITLE
Add tests non-implemented functions + loan bad arguments

### PR DESCRIPTION
**Pull request by Lobotuerk**
Signed-off-by: lobotuerk <jtlorente@ekumenlabs.com>
Related to https://github.com/ros2/rmw/pull/270

### CODE REVIEW

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:390]: @Lobotuerk if borrowing is successful, the message has to be explicitly returned or we'll leak.

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:385]: @Lobotuerk doesn't the `TestPublisherUse` fixture provide a `pub` already?

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:402]: @Lobotuerk hmm, if you're testing double borrowing, making sure the first call succeeded is important.

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:393]: @Lobotuerk meta: in general, after an error expectation, is important to reset the global error state using `rcl_reset_error()`.

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [reply test_rmw_implementation/test/test_publisher.cpp:402]: Also, returning the message is important too to not leak.

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:409]: @Lobotuerk meta: I see this a bad arguments test, for the most part as it has to check if ROS message loaning is at all supported. 

Considering that every single test involving ROS message loaning will have to perform the same check, why not splitting those checks out. For instance, having a fixture for ROS message loaning related tests that attempts a borrow on `SetUp()` and `GTEST_SKIP()` the whole thing if the operation is unsupported.

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:390]: By explicitly returning, do you mean with ```rmw_return_loaned_message_from_publisher```?

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:385]: Yes it does, thank you

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:398]: @Lobotuerk I think you want to do this right before failing the test.

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:393]: @Lobotuerk IIRC `GTEST_SKIP()` ends the test, so the `else` clause is redundant (i.e. not necessary).

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:410]: @Lobotuerk consider checking that `msg_pointer` remains `nullptr`. Same elsewhere.

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:418]: @Lobotuerk unused?

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:423]: @Lobotuerk why would you expect this to fail?

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:395]: @Lobotuerk meta: I think this should go into test cases that have not yet been implemented, if any. WDYT?

Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:395]: Yes, I was preparing  the template for future tests

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:418]: Should I instead check that the pointer is still null instead of assigning null again?

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:423]: Borrowing on top of an already burrowed msg? IE, pointer not null?

**Review by Blast545 — COMMENTED** [review]
Minor comments

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:390]: Doesn't seem to be one of those here

Comment (Blast545) [reply test_rmw_implementation/test/test_publisher.cpp:390]: `pub` is a `rmw_publisher_t`, so you have access to the boolean directly. Refer to this: [rmw_publisher_t](http://docs.ros2.org/latest/api/rmw/structrmw__publisher__t.html#a4c517b24ec0acd75690db63b9be43064)

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Blast545) [reply test_rmw_implementation/test/test_publisher.cpp:424]: That's not clear reading the code, you can add a comment line before the call to the `rmw_borrow_loaned_message`
```suggestion
  // Not null msg_pointer invalid to borrow message
  ret = rmw_borrow_loaned_message(pub, ts, &msg_pointer);
```

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Blast545) [reply test_rmw_implementation/test/test_publisher.cpp:390]: IMO my advice is a better approach, but I won't block the PR on that.

**Review by Blast545 — APPROVED** [review]
LGTM, remember to run CI on Tier 1 platforms before merging.

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:390]: Totally, but i did not know about that boolean.

**Review by Blast545 — APPROVED** [review]
LGTM

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:420]: Added

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:389]: @Lobotuerk oh, didn't know about this one! +1000

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:426]: @Lobotuerk you have to set `msg_pointer` back to `nullptr` here, nothing else will.

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:406]: @Lobotuerk nit^2: here and below, I'd tuck this expectation next to the previous call which is the one that could've (incorrectly) modified it.

**Review by hidmic — APPROVED** [review]
LGTM too, pending green CI

**Comment by hidmic** [comment]
Also, @Lobotuerk please consider changing the title so that it reads better -- once you squash, it'll become the commit message.

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:389]: I can't take the credit, @Blast545 told me about it

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp]: changes

File: test_rmw_implementation/test/test_publisher.cpp
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:407]: Changed

File: test_rmw_implementation/test/test_publisher.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_publisher.cpp:403]: @Lobotuerk hmm, it's unusual to test in `SetUp` and then `GTEST_SKIP` the rest. What happens if one of these expectations fails?

Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:403]: No idea, but I don't see another way of using both the variable and the functions returns

Comment (Lobotuerk) [reply test_rmw_implementation/test/test_publisher.cpp:403]: Update: If any of the `EXPECTS` triggers an error, the test does not get skipped and it goes on to the tests on those implementations, which the ones not implemented give Errors. IMO seems a fair outcome.

**Comment by hidmic** [comment]
@ros-pull-request-builder retest this please.

**Comment by Lobotuerk** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12393/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7361/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10104/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12306/)

**Comment by Lobotuerk** [comment]
All CI green, merging

**Pull request merged**
